packages/@aws-cdk/core/lib/private/tree-metadata.ts

```diff
@@ -20,37 +21,42 @@ export class TreeMetadata extends Construct {
   }
 
   protected synthesize(session: ISynthesisSession) {
-    const lookup: { [path: string]: Node } = { };
+    try {
+      const lookup: { [path: string]: Node } = { };
 
-    const visit = (construct: IConstruct): Node => {
-      const children = construct.node.children.map(visit);
-      const childrenMap = children.reduce((map, child) => Object.assign(map, { [child.id]: child }), {});
-      const node: Node = {
-        id: construct.node.id || 'App',
-        path: construct.node.path,
-        children: children.length === 0 ? undefined : childrenMap,
-        attributes: this.getAttributes(construct)
-      };
+      const visit = (construct: IConstruct): Node => {
+        const children = construct.node.children.map(visit);
+        const childrenMap = children.reduce((map, child) => Object.assign(map, { [child.id]: child }), {});
+        const node: Node = {
+          id: construct.node.id || 'App',
+          path: construct.node.path,
+          children: children.length === 0 ? undefined : childrenMap,
+          attributes: this.getAttributes(construct)
+        };
 
-      lookup[node.path] = node;
+        lookup[node.path] = node;
 
-      return node;
-    };
+        return node;
+      };
 
-    const tree = {
-      version: 'tree-0.1',
-      tree: visit(this.node.root),
-    };
+      const tree = {
+        version: 'tree-0.1',
+        tree: visit(this.node.root),
+      };
 
-    const builder = session.assembly;
-    fs.writeFileSync(path.join(builder.outdir, FILE_PATH), JSON.stringify(tree, undefined, 2), { encoding: 'utf-8' });
+      const builder = session.assembly;
+      fs.writeFileSync(path.join(builder.outdir, FILE_PATH), JSON.stringify(tree, undefined, 2), { encoding: 'utf-8' });
 
-    builder.addArtifact('Tree', {
-      type: ArtifactType.CDK_TREE,
-      properties: {
-        file: FILE_PATH
-      }
-    });
+      builder.addArtifact('Tree', {
+        type: ArtifactType.CDK_TREE,
+        properties: {
+          file: FILE_PATH
+        }
+      });
+    } catch (e) {
+      // Prevent errors here to fail synthesis, until this module stays @experimental.
+      this.node.addWarning(`tree.json could not be generated - ${e}`);
```

Comment:
This means that if any `visit` fails, the entire tree will not be synthed, which seems harsh. Wouldn't it make more sense to only surround the user code (call to `visit`) with this all-encompassing catch?

Reply:
I would rather provide no tree.json (which the consumer, an IDE for instance, can interpret as an error) rather than one that has missing information (and no indication that it is missing information).

Reply:
because of the extensible nature of this, I don't think it makes sense to fail the entire thing. Say you imported some 3rd party construct library that throw an exception in `inspect`. Suddenly, you can't see anything.

Maybe we can synthesize an "error" node or something if a node fails "visit", but I'd argue that failing the entire tree is not the right granularity.

Reply:
Fair enough. We could even collect and throw all the errors into the cloud manifest under the 'tree' metadata, which the consumer can collect and display.

Reply:
I think `.addWarning` should do the trick